File: LimSeoyeon/02_Queue/BOJ_3190_뱀.cpp
#include <iostream>
#include <queue>
using namespace std;

int arr[105][105];
int dy[4] = { 0,-1,0,1 };
int dx[4] = { 1,0,-1,0};

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	int N,M;
	cin >> N>>M;
	for (int i = 0; i < M; i++) {
		int a, b;
		cin >> a >> b;
		arr[a][b] = 1;
	}
	int orderNum;
	cin >> orderNum;

	queue<pair<int, int>> order;
	queue<pair<int, int>> q;

	for (int i = 0; i < orderNum; i++) {
		int a; 
		char b;
		cin >> a >> b;
		//오른쪽 회전 -> -, 왼쪽 회전-> +
		if (b == 'L') {
			order.push({ a, 1});
		}
		else {
			order.push({ a, -1 });
		}
	}
	int curY =1 ,curX = 1;
	int res = 0;
	int dirNum = 0;

	q.push({ 1,1 });

	while (true) {
		//1. 다음에 움직일 방향 설정
		if (!order.empty() && order.front().first == res) {
			dirNum += order.front().second;
			if (dirNum < 0) dirNum = 3;
			if (dirNum > 3) dirNum = 0;
			order.pop();
		}
		//2. 설정된 방향으로 다음에 움직일 위치 정하기
		curX += dx[dirNum];
		curY += dy[dirNum];

		//종료 조건 = 다음에 갈 위치가 벽이거나 자기자신의 몸이면 게임 종료.
		if (curY<1 || curX<1 || curY>N || curX>N || arr[curY][curX] == 2) {
			res += 1;
			break;
		}

		//3. 시간 흐르기
		res++;

		//4. 움직이기
		if (arr[curY][curX] == 0) { // 사과 없을 때 - 꼬리 없애기.
			int delY = q.front().first;
			int delX = q.front().second;
			arr[delY][delX] = 0;
			q.pop();
		}
		q.push({ curY,curX }); 
		arr[curY][curX] = 2;
	}

	cout << res;

}

File: LimSeoyeon/02_Queue/BOJ_3190_뱀.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

    public static class Pair{
        int first,second;
        Pair(int first, int second){
            this.first=first;
            this.second=second;
        }
    };

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer stk = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(stk.nextToken());
        stk = new StringTokenizer(br.readLine());
        int M = Integer.parseInt(stk.nextToken());

        int [][] arr = new int[N+5][N+5]; //보드판 배열
        int []dy = {0,-1,0,1}; //y축 방향 배열
        int []dx = {1,0,-1,0}; //x축 방향 배열

        for(int i=0; i<M; i++){
            stk = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(stk.nextToken());
            int b = Integer.parseInt(stk.nextToken());
            arr[a][b]=1; //사과 위치 1로 표시
        }
        stk = new StringTokenizer(br.readLine());
        int orderNum = Integer.parseInt(stk.nextToken());
        Queue<Pair> order = new LinkedList<>();
        Queue<Pair> q = new LinkedList<>();

        for(int i=0; i<orderNum; i++){
            stk = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(stk.nextToken());
            char b = stk.nextToken().charAt(0);
            if(b=='L'){
                order.add(new Pair(a,1));
            }
            else{
                order.add(new Pair(a,-1));
            }
        }

        int curY = 1, curX = 1;
        int res = 0; //시간 초
        int dirNum = 0;
        
        q.add(new Pair(1,1));

        while(true){
            //1. 다음에 움직일 방향 설정
            if(!order.isEmpty()&& order.peek().first==res){
                dirNum +=order.peek().second;
                if(dirNum<0) dirNum=3;
                if(dirNum>3) dirNum=0;
                order.poll();
            }

            curY+=dy[dirNum];
            curX+=dx[dirNum];

            if(curY<1|| curX<1 || curY >N || curX > N|| arr[curY][curX]==2){
                res+=1; //다음 초에 게임 종료됨. 지금은 움직이기 전이기 때문.
                break;
            }

            if(arr[curY][curX]==0) { //사과 없을 때 - 꼬리 없애기
                int delY = q.peek().first;
                int delX = q.peek().second;
                arr[delY][delX]=0;
                q.poll();
            }
            q.add(new Pair(curY,curX));
            arr[curY][curX]=2;

            res++;
        }

        System.out.println(res);
    }

}
